### PR TITLE
Avoid the binding operation when generating lambda byte-code

### PR DESCRIPTION
**Pull request by wenleix**
Work in progress. Binding lambda method to `this` is not fixed yet.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -982,18 +982,25 @@ protected Object visitLambdaExpression(LambdaExpression node, Object context)
...
+            List<Class<?>> argumentTypes =
+                    Stream.concat(
+                            Stream.of(ConnectorSession.class),
+                            functionType.getArgumentTypes().stream()
+                                    .map(Type::getJavaType)
+                                    .map(Primitives::wrap))
+                            .collect(toImmutableList());
+            List<String> argumentNames =
+                    Stream.concat(
+                            Stream.of("$connector_session"),
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Looks hacking to hard code this. But not sure if we should put a private static variable in this huge class.

**Comment by wenleix** [comment]
We can refactor `CachedInstanceBinder` into `InstanceFieldReigster`. Need to take some care of `this` variable in the same class.

**Comment by wenleix** [comment]
The original solution that refactors `CacheInstanceBinder` is backup in this branch: https://github.com/wenleix/presto/commits/fix_lambda_bind_backup2. Especially, the decoupling `CacheInstanceBinder` with `CallSiteBinder` is probably something want to refactor in a separate thread: https://github.com/wenleix/presto/commit/1bdc255c71eb6e4fdb05eb10cb8d2d650336142d

**Comment by wenleix** [comment]
Tested with the following query:

```
select count(transform(flatten(map_values(profile)), x -> x)) 
from ...
```

With the following VM argument enables:

```
-verbose:class
-Djava.lang.invoke.MethodHandle.DUMP_CLASS_FILES=true
```

Without the patch, it will generate about 23K  `LambdaForm$BMH*.class` files after processing 60K rows. With the patch, it generates ~420 files (there are about 300-350 files before the query start to run) after processing 75k rows.

**Comment by haozhun** [comment]
Use commit message `Minor fixes to ...`,  `Avoid binding session for every row for lambda execution`, `Avoid binding "this" for every row for lambda execution`

**Comment by haozhun** [comment]
* [ ] https://github.com/prestodb/presto/pull/7866/commits/fa661b5 Various small fixes to bytecode debug info
* [ ] https://github.com/prestodb/presto/pull/7866/commits/bfbcdcc Avoid binding session in lambda bytecode gen
* [ ] https://github.com/prestodb/presto/pull/7866/commits/b5bc8ef Avoid binding "this" in lambda bytecode gen

File: presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java [hunk 2]
```diff
@@ -87,7 +96,13 @@ public void generateMethods(ClassDefinition classDefinition, CallSiteBinder call
         constructorBody.comment("super();")
                 .append(thisVariable)
                 .invokeConstructor(Object.class);
+
         cachedInstanceBinder.generateInitializations(thisVariable, constructorBody);
+        for (PreGeneratedExpressions preGeneratedExpressions : allPreGeneratedExpressions) {
+            preGeneratedExpressions.getLambdaFieldMap().values().forEach(
+                    lambdaExpressionField -> lambdaExpressionField.generateInitialization(thisVariable, constructorBody)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/CursorProcessorCompiler.java]: Use a for-loop here instead. I don't have a particularly strong opinion. But I think it's easier to read that way.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java [hunk 3]
```diff
@@ -164,13 +171,21 @@ private static void generateConstructor(ClassDefinition classDefinition, FieldDe
 
         body.append(thisVariable.setField(sessionField, sessionParameter));
         cachedInstanceBinder.generateInitializations(thisVariable, body);
+        preGeneratedExpressions.getLambdaFieldMap().values().forEach(
+                lambdaExpressionField -> lambdaExpressionField.generateInitialization(thisVariable, body)
+        );
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java]: Use for-loop. It's definitely easier to read that way here.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 4]
```diff
@@ -531,6 +536,9 @@ private static void generateConstructor(
         additionalStatements.accept(constructorDefinition);
 
         cachedInstanceBinder.generateInitializations(thisVariable, body);
+        preGeneratedExpressions.getLambdaFieldMap().values().forEach(
+                lambdaExpressionField -> lambdaExpressionField.generateInitialization(thisVariable, body)
+        );
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Use for-loop

**Comment by wenleix** [comment]
This solves https://github.com/prestodb/presto/issues/7935 for un-captured lambda execution. Captured lambda execution needs to totally get rid of the usage of `MethodHandle`.

**Comment by wenleix** [comment]
The Jenkin failure after rebasing seems to be unrelated (EC2 issues? ) :

```
2017-05-15 00:33:46 INFO: FAILURE     /    com.facebook.presto.tests.hive.TestHiveStorageFormats.testInsertIntoPartitionedTable (Groups: storage_formats) took 5.3 seconds
2017-05-15 00:33:46 SEVERE: Failure cause:
com.teradata.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20170515_003341_00152_44wft): Error committing write to Hive
	at com.teradata.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:124)
	at com.teradata.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:85)
	at com.teradata.tempto.query.QueryExecutor.query(QueryExecutor.java:69)
	at com.facebook.presto.tests.hive.TestHiveStorageFormats.testInsertIntoPartitionedTable(TestHiveStorageFormats.java:171)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: Query failed (#20170515_003341_00152_44wft): Error committing write to Hive
	at com.facebook.presto.jdbc.PrestoResultSet.resultsException(PrestoResultSet.java:1799)
	at com.facebook.presto.jdbc.PrestoResultSet$ResultsPageIterator.computeNext(PrestoResultSet.java:1787)
	at com.facebook.presto.jdbc.PrestoResultSet$ResultsPageIterator.computeNext(PrestoResultSet.java:1756)
	at com.facebook.presto.jdbc.internal.guava.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:145)
	at com.facebook.presto.jdbc.internal.guava.collect.AbstractIterator.hasNext(AbstractIterator.java:140)
	at com.facebook.presto.jdbc.internal.guava.collect.TransformedIterator.hasNext(TransformedIterator.java:42)
	at com.facebook.presto.jdbc.internal.guava.collect.MultitransformedIterator.hasNext(MultitransformedIterator.java:50)
	at com.facebook.presto.jdbc.internal.guava.collect.MultitransformedIterator.hasNext(MultitransformedIterator.java:50)
	at com.facebook.presto.jdbc.PrestoResultSet.next(PrestoResultSet.java:149)
	at com.facebook.presto.jdbc.PrestoStatement.execute(PrestoStatement.java:222)
	at com.facebook.presto.jdbc.PrestoStatement.executeLargeUpdate(PrestoStatement.java:426)
	at com.facebook.presto.jdbc.PrestoStatement.executeUpdate(PrestoStatement.java:398)
	at com.teradata.tempto.query.JdbcQueryExecutor.executeQueryNoParams(JdbcQueryExecutor.java:137)
	at com.teradata.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:117)
	... 16 more
Caused by: com.facebook.presto.spi.PrestoException: Error committing write to Hive
	at com.facebook.presto.hive.RecordFileWriter.commit(RecordFileWriter.java:157)
	at com.facebook.presto.hive.HiveWriter.commit(HiveWriter.java:65)
	at com.facebook.presto.hive.HivePageSink.doFinish(HivePageSink.java:184)
	at com.facebook.presto.hive.authentication.UserGroupInformationUtils.lambda$executeActionInDoAs$0(UserGroupInformationUtils.java:29)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1678)
	at com.facebook.presto.hive.authentication.UserGroupInformationUtils.executeActionInDoAs(UserGroupInformationUtils.java:27)
	at com.facebook.presto.hive.authentication.ImpersonatingHdfsAuthentication.doAs(ImpersonatingHdfsAuthentication.java:39)
	at com.facebook.presto.hive.HdfsEnvironment.doAs(HdfsEnvironment.java:76)
	at com.facebook.presto.hive.HivePageSink.finish(HivePageSink.java:175)
	at com.facebook.presto.spi.classloader.ClassLoaderSafeConnectorPageSink.finish(ClassLoaderSafeConnectorPageSink.java:57)
	at com.facebook.presto.operator.TableWriterOperator.finish(TableWriterOperator.java:159)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:323)
	at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:234)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:537)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:229)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:623)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:458)
	... 3 more
Caused by: org.apache.hadoop.ipc.RemoteException: File /tmp/presto-hdfs/1812760f-1f5f-4271-905b-28d5a64b06ef/returnflag=N/20170515_003341_00152_44wft_a752f811-6cee-4e85-a859-f5edf15ffc42 could only be replicated to 0 nodes instead of minReplication (=1).  There are 1 datanode(s) running and no node(s) are excluded in this operation.
	at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.chooseTarget4NewBlock(BlockManager.java:1622)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getAdditionalBlock(FSNamesystem.java:3351)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.addBlock(NameNodeRpcServer.java:683)
	at org.apache.hadoop.hdfs.server.namenode.AuthorizationProviderProxyClientProtocol.addBlock(AuthorizationProviderProxyClientProtocol.java:214)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.addBlock(ClientNamenodeProtocolServerSideTranslatorPB.java:495)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:617)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1073)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2141)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2137)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1714)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2135)

	at org.apache.hadoop.ipc.Client.call(Client.java:1475)
	at org.apache.hadoop.ipc.Client.call(Client.java:1412)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:229)
	at com.sun.proxy.$Proxy251.addBlock(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.addBlock(ClientNamenodeProtocolTranslatorPB.java:418)
	at sun.reflect.GeneratedMethodAccessor606.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
	at com.sun.proxy.$Proxy252.addBlock(Unknown Source)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.locateFollowingBlock(DFSOutputStream.java:1455)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.nextBlockOutputStream(DFSOutputStream.java:1251)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.run(DFSOutputStream.java:448)
```

**Pull request merged**